hspec-discover/src/Run.hs

```diff
@@ -119,15 +119,31 @@ findSpecs src = do
   mapMaybe (fileToSpec dir) . filter (/= file) <$> getFilesRecursive dir
 
 fileToSpec :: FilePath -> FilePath -> Maybe Spec
-fileToSpec dir file = case reverse $ splitDirectories file of
-  x:xs -> case stripSuffix "Spec.hs" x <|> stripSuffix "Spec.lhs" x of
-    Just name | (not . null) name -> Just . Spec (dir </> file) $ (intercalate "." . reverse) (name : xs)
-    _ -> Nothing
-  _ -> Nothing
+fileToSpec dir sfile = do
+    file@(_:_) <- stripSuffix "Spec.hs" sfile <|>
+                  stripSuffix "Spec.lhs" sfile
+    mname <- getModuleName file
+    return $ Spec (dir </> sfile) mname
   where
     stripSuffix :: Eq a => [a] -> [a] -> Maybe [a]
     stripSuffix suffix str = reverse <$> stripPrefix (reverse suffix) (reverse str)
 
+getModuleName :: FilePath -> Maybe String
+getModuleName f
+    | last f == pathSeparator = Nothing
```

Comment:
`last` is partial.  Even though we checked for `null` above, I still think a solution that uses pattern matching is preferable.